lua/apisix/plugins/basic-auth.lua

```diff
@@ -0,0 +1,225 @@
+    core.log.info("plugin access phase, conf: ", core.json.delay_encode(conf))
+
+    -- 0. check the plugin is enabled
+    if not conf.enable then
+        return
+    end
+
+
+    -- 1. extract authorization from header
+    local headers = ngx.req.get_headers()
+    if not headers.Authorization then
+        return 401, { message = "authorization is required" }
+    end
+
+    local username, password, err = extract_auth_header(headers.Authorization)
+    if err then
+        return 401, { message = err }
+    end
+
+    -- 2. get user info from etcd
+    local res = authorizations_etcd:get(username)
+    if res == nil then
+        return 401, { message = "failed to find authorization from etcd" }
+    end
+
+    -- 3. check user exists
+    if not res.value or not res.value.id then
+        return 401, { message = "user is not found" }
+    end
+
+    local value = res.value
+    core.log.info("etcd value: ", core.json.delay_encode(value))
+
+    -- 4. check the password is correct
+    if value.password ~= password then
+        return 401, { message = "password is error" }
+    end
+
+
+    core.log.info("hit basic-auth access")
+end
+
+
+local function set_auth()
+    local body_table = {}
+    -- read_body can not use in log_by_lua
+    if ngx.re.find(ngx.req.get_headers()["Content-Type"] or "", "application/json") then
+        ngx.req.read_body()
+
+        local body_data = ngx.req.get_body_data()
+        if body_data ~= nil then
+            body_table = json.decode(body_data)
+        end
+
+    else
+        body_table = ngx.req.get_post_args()
+    end
+
+    local username = body_table["username"]
+    local password = body_table["password"]
+
+    if not username or not password then
+        core.response.exit(200, "username,password is required")
+    end
+
+    local key = gen_key(username)
+
+    local res, err = core.etcd.set(key, { username = username, password = password })
+    if not res then
+        core.response.exit(500, err)
+    end
+
+    core.response.exit(res.status, res.body)
+end
+
+local function get_auth()
+    local request_table = ngx.req.get_uri_args() or {}
+
+    if not request_table["username"] then
+        core.response.exit(200, "username is required")
+    end
+
+    local username = request_table["username"]
+
+    local key = gen_key(username)
+
+    local res, err = core.etcd.get(key)
+    if not res then
+        core.response.exit(500, err)
+    end
+
+    core.response.exit(res.status, res.body)
+end
+
+-- curl 'http://127.0.0.1:9080/apisix/plugin/basic-auth/set' -H "Content-Type:application/json" -d '{"username":"foo","password":"bar"}'
+
+function _M.api()
+    return {
+        {
+            methods = { "GET" },
+            uri = "/apisix/plugin/basic-auth/get",
+            handler = get_auth,
+        },
+        {
+            methods = { "POST", "PUT" },
+            uri = "/apisix/plugin/basic-auth/set",
+            handler = set_auth,
+        }
+    }
+end
```

Comment:
I don't think we need export admin API in this way.

Reply:
The `POST`, `PUT`, `DELETE` admin APIs will be deal with the core of APISIX, plugin don't need to care about those things.